Sort and relevance cut-off Part 2

wanliAlex opened the pull request:
## Change Summary
<!-- Describe the key changes this PR introduces -->
This is the second part of the sort feature that includes the implementation of the sort functionality. 
This PR includes the following changes:
This pull request introduces enhancements to the hybrid search functionality, focusing on implementing support for sorting by fields (`sortBy`) and refining post-processing logic. Below are the key changes grouped by theme.

### Hybrid Search Enhancements:
* Introduced a new `SortField` class to define field-specific sorting configurations (`field_name`, `order`, `missing`) in `HybridSearcher` (`vespa/src/main/java/ai/marqo/search/HybridSearcher.java`).
* Added support for `sortBy` parameters (`fields`, `sortDepth`, `sortCandidates`) in the query processing pipeline. These parameters are used to configure sorting behavior during hybrid search (`vespa/src/main/java/ai/marqo/search/HybridSearcher.java`) [[1]](diffhunk://#diff-da6475dc2a9b7d803bba658add780db7404078c715c3e13d2cd93ad95b2f4077R81-R89) [[2]](diffhunk://#diff-da6475dc2a9b7d803bba658add780db7404078c715c3e13d2cd93ad95b2f4077R132-R137).
* Implemented a `postProcessBySort` method to sort search hits based on specified `sortBy` fields, with support for null handling and relevance as a tie-breaker. This method also resets relevance scores to reflect the new order (`vespa/src/main/java/ai/marqo/search/HybridSearcher.java`).
* Updated the main search logic to conditionally apply sorting or global score modifiers based on query parameters, ensuring compatibility with existing features (`vespa/src/main/java/ai/marqo/search/HybridSearcher.java`).

## Related Jira Ticket
<!-- Link to Jira ticket (e.g. MAR-123) -->

## Checklist
<!-- Check items that apply, add items if needed -->
- [*] Tests have been added for changes
- [n/a] Documentation has been updated
- [n/a] Breaking changes are clearly identified
- [n/a] Python client changes linked or N/A

<!-- Special test requirements for certain changes -->
### For new field types:
- [ ] Tests cover score modifier usage of this new type
- [ ] Test indexes updated to cover the new type for all APIs (add docs, search, partial update, etc.)

wanliAlex committed:
Finish API implementation

wanliAlex committed:
Finish unit tests API for sort

wanliAlex committed:
unit tests for query generation

wanliAlex committed:
Add single field tests for java

wanliAlex committed:
Adding integ tests now

wanliAlex committed:
Need to solve inconsistent results issue

wanliAlex committed:
Add more tests

wanliAlex committed:
Add tests for 3 fields

wanliAlex committed:
Add tests for 3 fields

wanliAlex committed:
Add tests for 3 fields

wanliAlex committed:
Fix schemas

wanliAlex committed:
Merge branch 'mainline' into li/sort-and-relevance-cuffoff-python-api

wanliAlex committed:
Revert the vespa change

wanliAlex committed:
delete the integration tests for sort by

wanliAlex committed:
Fix tests

wanliAlex committed:
Fix tests

wanliAlex committed:
Add tests for sort

wanliAlex committed:
Add tests for sort

wanliAlex committed:
Add tests for sort

wanliAlex committed:
Restore pre-216 schemas

wanliAlex committed:
Fix tests

wanliAlex committed:
Fix tests

wanliAlex committed:
Fix tests

wanliAlex committed:
Fix tests

wanliAlex committed:
Fix tests

wanliAlex committed:
Finish Vespa implementation and add tests

wanliAlex committed:
Catch mainline

wanliAlex committed:
Add default tests

wanliAlex committed:
Fix tests

wanliAlex committed:
Fix tests

wanliAlex committed:
Fix tests

wanliAlex committed:
Fix java part

wanliAlex committed:
Add index type tests

wanliAlex committed:
Fix java code

wanliAlex committed:
Add tests for java

wanliAlex committed:
Fix python tests

wanliAlex committed:
Add tests for java

wanliAlex committed:
Add java tests

wanliAlex committed:
Fix typo

wanliAlex committed:
Fix typo

wanliAlex committed:
Add API tests

wanliAlex committed:
Fix tests

papa99do commented on vespa/src/main/java/ai/marqo/search/HybridSearcher.java:
```diff
@@ -184,10 +204,33 @@ public Result search(Query query, Execution execution) {
+                    postProcessBySort(
+                            hitsForPostProcessing, sortByFields, sortBySortDepth, limit, offset);
+            processedHits.setField("marqo__sortCandidates", hitsForPostProcessing.size());
+        } else if ((queryMultWeightsGlobal != null && !queryMultWeightsGlobal.isEmpty())
+                || (queryAddWeightsGlobal != null && !queryAddWeightsGlobal.isEmpty())) {
```
This logic does not match the previous one. Why do we need to add this branch here?

wanliAlex replied:
Revert the changes. We use the old postProcess when sort is not provided.

papa99do commented on vespa/src/main/java/ai/marqo/search/HybridSearcher.java:
```diff
@@ -230,6 +273,81 @@ public Result search(Query query, Execution execution) {
+            Integer limit,
+            Integer offset) {
+
+        ObjectMapper mapper = new ObjectMapper();
```
There's better way to use jackson's thread-safe ObjectReader, see https://stackoverflow.com/questions/35335646/how-correctly-and-thread-safe-reuse-jackson-objectreader for details. We can declare the ObjectReader as a instance property.

wanliAlex replied:
done

papa99do commented on vespa/src/main/java/ai/marqo/search/HybridSearcher.java:
```diff
@@ -230,6 +273,81 @@ public Result search(Query query, Execution execution) {
+        try {
+            parsedSortByFields =
+                    mapper.readValue(sortByFields, new TypeReference<List<SortField>>() {});
+        } catch (Exception e) {
```
Be more specific about what exception we catch. Here, we probably just want to catch IOException.

wanliAlex replied:
done

papa99do commented on vespa/src/main/java/ai/marqo/search/HybridSearcher.java:
```diff
@@ -230,6 +273,81 @@ public Result search(Query query, Execution execution) {
+            parsedSortByFields =
+                    mapper.readValue(sortByFields, new TypeReference<List<SortField>>() {});
+        } catch (Exception e) {
+            throw new RuntimeException(
```
nit. we'll need to change the way we handle exception in CustomSearcher. Ideally, we should define some specific exceptions instead of just a general runtime exception. This is similar to throw an InternalError for everything in python app, which is not a good practice. We can leave this to a follow-up refactoring. 
RuntimeException will cause a 500 error response of Vespa, is this OK? We have validation logic in Marqo, right?

papa99do commented on vespa/src/main/java/ai/marqo/search/HybridSearcher.java:
```diff
@@ -115,6 +129,12 @@ public Result search(Query query, Execution execution) {
         // --- End facets subquery handling ---
 
+        // --- Update the query limit if sort is used
+        if (sortByFields != null && !sortByFields.isEmpty()) {
```
nit. there's a helper method you can use to replace this: `!Strings.isNullOrEmpty(sortByFields)`

wanliAlex replied:
done

papa99do commented on vespa/src/main/java/ai/marqo/search/HybridSearcher.java:
```diff
@@ -230,6 +273,81 @@ public Result search(Query query, Execution execution) {
+                        : new ArrayList<>();
+
+        // Build comparator chain based on configured SortFields
+        Comparator<Hit> sortComparator = null;
+        for (int i = 0; i < parsedSortByFields.size(); i++) {
+            final int idx = i;
+            SortField sf = parsedSortByFields.get(i);
+            Function<Hit, Double> keyExtractor =
+                    hit -> {
+                        FeatureData mf = (FeatureData) hit.getField("matchfeatures");
+                        if (mf == null) return null;
+                        double v = mf.getDouble("sort_field_value_" + idx);
+                        return (v == -1e50) ? null : v;
+                    };
+            Comparator<Double> base = Comparator.naturalOrder();
+            if ("desc".equalsIgnoreCase(sf.order)) {
+                base = base.reversed();
+            }
+            Comparator<Double> nullAware =
+                    "last".equalsIgnoreCase(sf.missing)
+                            ? Comparator.nullsLast(base)
+                            : Comparator.nullsFirst(base);
+            Comparator<Hit> fieldComparator = Comparator.comparing(keyExtractor, nullAware);
+            sortComparator =
+                    (sortComparator == null)
+                            ? fieldComparator
+                            : sortComparator.thenComparing(fieldComparator);
+        }
+
+        // Append relevance as final tie-breaker (always descending)
+        Comparator<Hit> relevanceTie =
+                Comparator.comparingDouble((Hit h) -> h.getRelevance().getScore()).reversed();
+        // always combine with relevance tie-break
+        sortComparator = sortComparator.thenComparing(relevanceTie);
```
This piece of logic can be extracted to an  HitOrderer implementation. This might help with the unit testing.  see `HitSortOrderer` as an example.

papa99do replied:
Ignore this. I realised that we are only sorting until the SortDepth. We could implement an orderer that does sort until depth, but leave it to the refactoring.

wanliAlex replied:
added todo.

papa99do commented on vespa/src/main/java/ai/marqo/search/HybridSearcher.java:
```diff
@@ -230,6 +273,81 @@ public Result search(Query query, Execution execution) {
+                        FeatureData mf = (FeatureData) hit.getField("matchfeatures");
+                        if (mf == null) return null;
+                        double v = mf.getDouble("sort_field_value_" + idx);
+                        return (v == -1e50) ? null : v;
```
`-1e50` is a magic number, extract it as a constant

wanliAlex replied:
done

papa99do commented on vespa/src/test/java/ai/marqo/search/HybridSearcherTest.java:
```diff
@@ -620,4 +623,401 @@ void shouldCreateProperFacetQueries() {
+    /**
+     * Test for sortBy feature in HybridSearcher.
+     */
+    @Nested
```
I think it's OK to have this test in a separate file given it's size.  This is the test cases I think we need to cover for postProcessBySort method (can leverage parameterised tests):

- [x]  sortJson parsing logic - happy path - with 1, 2, 3 sort fields, various missing and order value)
- [x] sortJson parsing logic - corner case - parsing error (IOException)
- [x] sortJson parsing logic - corner case - 0, or 3+ sort fields
- [x] sortJson parsing logic - corner case - unsupported missing or order value

- [x] comparator - happy path - 1 sort field, test last/first, asc/desc combination
- [x] comparator - happy path - 2 sort fields, with first field fixed, test last/first, asc/desc combination of the second field
- [x] comparator - happy path - 3 sort fields, with first two fields fixed, test last/first, asc/desc combination of the third field
- [x] comparator - happy path - 3 sort fields, test the relevance tie breaker. 
- [ ] comparator - corner case - missing field in matchfeatures (is this possible?)

- [x] sortBySortDepth - null/smaller than hit size/larger than hit size/equal to hit size
- [x] sortBySortDepth - corner case, 0 / less than 0 (might not need since python has validated this, but we probably should assert) 

- [x] offset and limit - make sure the pagination works properly, only the trimmed result are returned in the hit group.

wanliAlex replied:
Added tests for sortJson Parsing.

wanliAlex replied:
SortBySortDepth can be null, and all the retrievals will be sorted. Add a test to ensure it must be greater than or equal to 1.

wanliAlex replied:
> comparator - corner case - missing field in matchfeatures (is this possible?)

 If the field is missing, we return the magic number.

papa99do commented on vespa/src/main/java/ai/marqo/search/HybridSearcher.java:
```diff
@@ -47,6 +46,12 @@ public class HybridSearcher extends Searcher {
     private static String MARQO_SEARCH_METHOD_TENSOR = "tensor";
     private List<String> STANDARD_SEARCH_TYPES = new ArrayList<>();
 
+    private static class SortField {
+        public String field_name;
+        public String order;
+        public String missing;
+    }
+
```
- Consider using java record. 
> Records are immutable data classes that require only the type and name of fields.
- Consider using Enum for order and missing fields to make sure only supported value are parsed
- Seems the field_name field isn't used anywhere in the code. Is this expected? 
- If we'd like to keep field_name, consider using camel case `fieldName` with an jackson annotation: `@JsonProperty("field_name")`

wanliAlex replied:
done

papa99do commented on tests/integ_tests/tensor_search/integ_tests/test_search_sort_by_feature.py:
```diff
@@ -1,10 +1,142 @@
+from marqo.core.models.marqo_index import MarqoIndex
+
+
+class TestSearchSortByFeature(MarqoTestCase):
```
nit. since this is an integration test, we can use real index here. just specify the marqo_version when you create it.

wanliAlex replied:
I lean towards using the mock instead of creating real indexes, as I feel this is more reliable and modifying the index version is a bit hacky, especially for legacy unstructured indexes. Let me know your thoughts.

wanliAlex committed:
Fix Jackson ObjectReader thread-safety, IOException handling, and magic number

- Add thread-safe ObjectReader instance for SortField parsing
- Replace generic Exception catch with specific IOException
- Extract magic number -1e50 as MISSING_SORT_VALUE_SENTINEL constant

These changes address code review feedback for better thread safety
and code maintainability.

wanliAlex committed:
Fix the logic branch comment

wanliAlex committed:
Move SortBy Test to a file

wanliAlex committed:
Added Json parsing tests

wanliAlex committed:
Add json parsing tests

wanliAlex committed:
Catch JsonProcessingException

wanliAlex committed:
Added TODO

wanliAlex committed:
Added tests for sort integration tests

wanliAlex committed:
Added tests for sortBySortDepth

wanliAlex committed:
Add backwards compatibility tests

wanliAlex committed:
Add sort by feature

wanliAlex committed:
Fix backwards compatibility tests

wanliAlex committed:
Fix tests

wanliAlex commented on vespa/src/main/java/ai/marqo/search/HybridSearcher.java:
```diff
@@ -47,6 +51,47 @@ public class HybridSearcher extends Searcher {
+    private static final ObjectReader SORT_FIELD_READER =
+            OBJECT_MAPPER.readerFor(new TypeReference<List<SortField>>() {});
+
+    // Magic number for missing sort field values
+    private static final double MISSING_SORT_VALUE_SENTINEL = -1e50;
```
Done

wanliAlex commented on vespa/src/test/java/ai/marqo/search/HybridSearcherTest.java:
```diff
@@ -3,17 +3,19 @@
 import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
Done

wanliAlex committed:
Reponse to github copilot reviews

wanliAlex committed:
Fix the format issue

wanliAlex committed:
Fix a backwards compatibility test issue

papa99do commented on tests/compatibility_tests/conftest.py:
```diff
@@ -27,7 +27,7 @@ def pytest_collection_modifyitems(config, items):
                 item.add_marker(pytest.mark.skip(reason=f"Testcase: {item.name} marked with skip_marqo_version: {skip_marqo_versions}. Skipping."))
                 continue
-        elif test_case_version_marker:
+        if test_case_version_marker:
```
is there any difference?

wanliAlex replied:
This fixes a bug that if `skip_marqo_version` exists, the` test_case_version_marker` check is skipped.

papa99do commented on tests/compatibility_tests/search/unstructured/test_hybrid_search_unstructured_sort_by_feature.py:
```diff
@@ -0,0 +1,213 @@
+
+from tests.compatibility_tests.base_test_case.base_compatibility_test import BaseCompatibilityTestCase
+
+@pytest.mark.marqo_version('2.11.0')
```
2.21.0 or 2.16.0? 
Will this feature work for all indexes using the 2_16 schema template?

wanliAlex replied:
This marker actually means the `minimum_marqo_version` to test on. It means this test will run on any version greater than or equal to `2.11.0`

papa99do commented on tests/integ_tests/tensor_search/integ_tests/test_search_sort_by_feature.py:
```diff
@@ -743,4 +895,150 @@ def test_sort_by_when_fields_do_not_exist_three_fields(self):
+            # All documents before position of doc 7 should have the sort field
+            doc_id = ids[i]
+            # Documents 0,1,2,3,4,5,6,8,9 should all have sort_field_3 and come before doc 7
+            self.assertIn(doc_id, ['0', '1', '2', '3', '4', '5', '6', '8', '9'])
```
This does not actually test all ids are before doc_7, it actually does not test anything. I think the previous assertion is good enough to test 7 is the last. 
Also , can we add another doc with missing field3 to test tie breaking for missing fields?

wanliAlex replied:
1. Made doc6 also missing field 3, and added a test on this relevance tie between doc 7 and doc 6.
2. Added a test for a sort on a non-existent field and ensured all results are sorted by relevance.

papa99do commented on vespa/src/main/java/ai/marqo/search/HybridSearcher.java:
```diff
@@ -46,10 +51,47 @@ public class HybridSearcher extends Searcher {
+        @JsonCreator
+        public static SortOrder fromString(String value) {
+            if (value == null) return ASC; // default
+            return "desc".equalsIgnoreCase(value) ? DESC : ASC;
```
A good practice of handling input is `validate input rigorously and fail fast on bad data.` 
So it would be better to let invalid input error out instead of falling back to default. 
`return SortOrder.valueOf(value.toUpperCase(Locale.ROOT));` . This will throw IllegalArgumentException if a value other than `desc` or `asc` is provided.

wanliAlex replied:
Thanks! Made the change. We don't have to catch `IllegalArgumentException` as it will caught later and converted to RuntimeError

papa99do commented on vespa/src/test/java/ai/marqo/search/SortByTest.java:
```diff
@@ -0,0 +1,746 @@
+     * Test that verifies sorting of results based on a single sort field.
+     */
+    HitGroup helpGenerateHitGroupWithOnlySortFieldValue0() {
+        FeatureData f1 = mock(FeatureData.class);
+        when(f1.getDouble("sort_field_value_0")).thenReturn(-1e50);
+        Hit doc1 = new Hit("doc1", 0.55);
+        doc1.setField("matchfeatures", f1);
```
nit. consider extract a method to reduce dup code

papa99do commented on vespa/src/test/java/ai/marqo/search/SortByTest.java:
```diff
@@ -0,0 +1,746 @@
+        }
+
+        @Test
+        void shouldThrowExceptionForNullOrderAndMissing() {
```
is null is provided, we'll just use the default, right?

papa99do commented on vespa/src/test/java/ai/marqo/search/SortByTest.java:
```diff
@@ -0,0 +1,746 @@
+                                + order
+                                + "\",\"missing\":\"last\"}]";
+
+                // Should not throw exception, should default to ASC
```
this is not a good practice. we should validate the input rigorously and fail fast

wanliAlex replied:
Revised this test

papa99do commented on vespa/src/test/java/ai/marqo/search/SortByTest.java:
```diff
@@ -0,0 +1,746 @@
+                                + missing
+                                + "\"}]";
+
+                // Should not throw exception, should default to FIRST
```
same here

wanliAlex replied:
Revised this test

papa99do commented on vespa/src/test/java/ai/marqo/search/SortByTest.java:
```diff
@@ -0,0 +1,746 @@
+        }
+
+        @Test
+        void shouldValidateFieldsInCorrectOrder() {
```
not sure if this test case tests what it claims to test :D . It looks like it's still testing missing fields.

wanliAlex replied:
updated the test name to `shouldThrowWhenSortFieldJsonIsMissingOrderOrFieldNameAtCorrectIndex`

wanliAlex committed:
Respond to Yihan's comments

papa99do reviewed (dismissed):
LGTM

wanliAlex committed:
Fix format
